docs/example-iam-policy-strict.json

```diff
@@ -0,0 +1,126 @@
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:AttachVolume",
+        "ec2:DetachVolume",
+        "ec2:ModifyVolume",
```

Comment:
For these actions can we scope them further to only allow this on the volumes provisioned by the CSI driver?

Reply:
I _think_ I tried that originally, but ran into issues; I don't remember what exactly. If someone can confirm scoping this further won't cause any issue, I'm happy to do it

Reply:
Here is a cleaned up policy to do this. I also removed the broad `ec2:DescribeTags` action because there is no reference to that API call anywhere in the codebase. I wonder why it's there in the first place?

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "ec2:DescribeAvailabilityZones",
        "ec2:DescribeInstances",
        "ec2:DescribeSnapshots",
        "ec2:DescribeVolumes",
        "ec2:DescribeVolumesModifications"
      ],
      "Resource": "*"
    },
    {
      "Effect": "Allow",
      "Action": [
        "ec2:CreateTags"
      ],
      "Resource": [
        "arn:aws:ec2:*:*:volume/*",
        "arn:aws:ec2:*:*:snapshot/*"
      ],
      "Condition": {
        "StringEquals": {
          "ec2:CreateAction": [
            "CreateVolume",
            "CreateSnapshot"
          ]
        }
      }
    },
    {
      "Effect": "Allow",
      "Action": [
        "ec2:DeleteTags"
      ],
      "Resource": [
        "arn:aws:ec2:*:*:volume/*"
      ],
      "Condition": {
        "StringLike": {
          "ec2:ResourceTag/CSIVolumeName": "*",
          "ec2:ResourceTag/ebs.csi.aws.com/cluster": "true",
          "ec2:ResourceTag/kubernetes.io/cluster/<%= cluster_name %>": "owned"
        }
      }
    },
    {
      "Effect": "Allow",
      "Action": [
        "ec2:DeleteTags"
      ],
      "Resource": [
        "arn:aws:ec2:*:*:snapshot/*"
      ],
      "Condition": {
        "StringLike": {
          "ec2:ResourceTag/CSIVolumeSnapshotName": "*",
          "ec2:ResourceTag/ebs.csi.aws.com/cluster": "true",
          "ec2:ResourceTag/kubernetes.io/cluster/<%= cluster_name %>": "owned"
        }
      }
    },
    {
      "Effect": "Allow",
      "Action": [
        "ec2:CreateVolume"
      ],
      "Resource": "*",
      "Condition": {
        "StringLike": {
          "aws:RequestTag/CSIVolumeName": "*",
          "aws:RequestTag/ebs.csi.aws.com/cluster": "true",
          "aws:RequestTag/kubernetes.io/cluster/<%= cluster_name %>": "owned"
        }
      }
    },
    {
      "Effect": "Allow",
      "Action": [
        "ec2:AttachVolume",
        "ec2:DetachVolume"
      ],
      "Resource": [
        "arn:aws:ec2:*:*:volume/*"
      ],
      "Condition": {
        "StringLike": {
          "ec2:ResourceTag/CSIVolumeName": "*",
          "ec2:ResourceTag/ebs.csi.aws.com/cluster": "true",
          "ec2:ResourceTag/kubernetes.io/cluster/<%= cluster_name %>": "owned"
        }
      }
    },
    {
      "Effect": "Allow",
      "Action": [
        "ec2:AttachVolume",
        "ec2:DetachVolume"
      ],
      "Resource": [
        "arn:aws:ec2:*:*:instance/*"
      ],
      "Condition": {
        "StringLike": {
          "ec2:ResourceTag/kubernetes.io/cluster/<%= cluster_name %>": "owned"
        }
      }
    },
    {
      "Effect": "Allow",
      "Action": [
        "ec2:ModifyVolume",
        "ec2:DeleteVolume"
      ],
      "Resource": "*",
      "Condition": {
        "StringLike": {
          "ec2:ResourceTag/CSIVolumeName": "*",
          "ec2:ResourceTag/ebs.csi.aws.com/cluster": "true",
          "ec2:ResourceTag/kubernetes.io/cluster/<%= cluster_name %>": "owned"
        }
      }
    },
    {
      "Effect": "Allow",
      "Action": [
        "ec2:CreateSnapshot",
        "ec2:DeleteSnapshot"
      ],
      "Resource": "*",
      "Condition": {
        "StringLike": {
          "ec2:ResourceTag/CSIVolumeSnapshotName": "*",
          "ec2:ResourceTag/ebs.csi.aws.com/cluster": "true",
          "ec2:ResourceTag/kubernetes.io/cluster/<%= cluster_name %>": "owned"
        }
      }
    }
  ]
}

```

Reply:
Have you tested this? 

I don't think `CreateSnapshot` and `DeleteSnapshot` can be bundled together. The condition for `CreateSnapshot` would be `"aws:RequestTag...`. 

However, I remember running into issues trying to make this policy as clean as possible. Changes that seemed innocuous actually weren't.

Reply:
FWIW I did a basic test with this policy and it appears to work.

```
kubectl apply -f aws-ebs-csi-driver/examples/kubernetes/dynamic-provisioning/manifests
kubectl delete -f aws-ebs-csi-driver/examples/kubernetes/dynamic-provisioning/manifests
```

The trick is you need to scope the condition on the `instance` side and `volume` side in separate statements. If you do it on `*`, the instances will not have the additional tags that the volumes have, denying the driver permission to attach/detatch. Found it from this blog post [here](https://www.iampulse.com/resources/policies/attach-or-detach-volumes-to-an-ec2-instance-based-on-tags).

Reply:
Oh for the snapshots I did not. Good catch.

Reply:
Ok, so you tested this last policy and everything works? If so, I'll commit that

Reply:
I did not test snapshots yet. Are you able to test? Otherwise I can try to do it tomorrow.

Reply:
Ok, I tested this final [policy](https://gist.github.com/mitchellhuang/ed141207493745ea22ac02f27396df16) with snapshots and it appears to work. Ended up just doing `*` resources for `CreateSnapshot` like you had.